### PR TITLE
Implement model types guidelines

### PR DESCRIPTION
**Pull request by bterlson**
We don't really use the `details` convention but it's in the general guidelines so presumably at one point we all agreed to it? Regardless I made it optional. Otherwise these are just TS-specific tweaks.

### CODE REVIEW

File: docs/typescript/design.md [hunk 1]
```diff
@@ -608,9 +608,48 @@ TODO: If this is largely implemented for the API Designer, please include an exa
 
 TODO: Please add a section on conditional request methods.
 
-TODO: Please add a section providing guidance on model types.
+## Model Types {#ts-model-types}
 
-TODO: Please add discussion similar to the [Model Type discussion from the General Guidelines](https://azure.github.io/azure-sdk/general_design.html#model-types), including the naming table if relevant to JS/TS, or an alternate one specific to JS/TS.  
+Client libraries represent entities transferred to and from Azure services as model types. Certain types are used for round-trips to the service. They can be sent to the service (as an addition or update operation) and retrieved from the service (as a get operation). These must be named according to the type. For example, a `ConfigurationSetting` in App Configuration, or an `Event` on Event Grid.
+
+{% include requirement/MUST id="ts-model-types-use-good-name" %} follow the above convention for types which round-trip to the service and represent a complete entity.
```
Comment (annelo-msft) [inline docs/typescript/design.md:615]: Which above convention?  (Sorry I can't find it)

Comment (bterlson) [reply docs/typescript/design.md:615]: the line 613 content is what I was referring to

File: docs/typescript/design.md [hunk 2]
```diff
@@ -608,9 +608,48 @@ TODO: If this is largely implemented for the API Designer, please include an exa
...
+
+The following table enumerates the various models you might create:
+
+| Type | Example | Usage |
+| `<model>` | `Secret` | The full data for a resource |
+| `<model>Details` | `SecretDetails` | Less important details about a resource. Attached to `<model>.details` |
+| `<model>Item` | `SecretItem` | A partial set of data returned for enumeration |
+| `<operation>Options` | `AddSecretOptions` | Optional parameters to a single operation |
+| `<operation>Result` | `AddSecretResult` | A partial or different set of data for a single operation |
+| `<model><verb>Result` | `SecretChangeResult` | A partial or different set of data for multiple operations on a model |
```
Comment (annelo-msft) [inline docs/typescript/design.md:652]: Do you follow these conventions in JS?  Just want to make sure I'm not making you add guidelines that people won't find followed in Track 2 SDKs.

**Review by annelo-msft — APPROVED** [review]

File: docs/typescript/design.md (hunk 2 above)
Comment (bterlson) [reply docs/typescript/design.md:652]: I don't think we have any *Results now, but if we had them, it seems like a fine name.

File: docs/typescript/design.md (hunk 1 above)
Comment (annelo-msft) [reply docs/typescript/design.md:615]: Could you sharpen the language to be more clear in the guideline regarding what convention you're asking them to follow?  Is it naming them according to the type?

Comment (bterlson) [reply docs/typescript/design.md:615]: Yes. I'll repeat that verbiage!

Comment (annelo-msft) [reply docs/typescript/design.md:615]: Sorry I am dense. It just kind of got lost in the discussion there for me.

**Review by xirzec — APPROVED** [review]
I like these changes a lot, feels like an early holiday gift! 

One minor concern, not sure if it's just GitHub's markdown previewer, but the table didn't seem to preview correctly at least. I hope it renders on the actual site?

**Comment by bterlson** [comment]
I confirmed locally and it seems to render fine... I'll take a peek once this is merged to make sure the live site looks ok.

**Comment by bterlson** [comment]
Forgot to merge this a while back, but still seems good to go!

**Pull request merged**